File: src/main/java/xyz/realmath/AbelianGroup.java
package xyz.realmath;

public interface AbelianGroup<T extends AbelianGroup<T>> {
  default boolean isZero() {
    return this.equals(this.add(thisT()));
  }

  T add(T t);

  T negate();

  default T subtract(T t) {
    return thisT().add(t.negate());
  }

  @SuppressWarnings("unchecked")
  default T thisT() {
    return (T) this;
  }
}

File: src/main/kotlin/xyz/realmath/AbelianGroup.kt
package xyz.realmath

interface AbelianGroup<T : AbelianGroup<T>> {

    operator fun plus(t: T): T

    operator fun unaryMinus(): T

    fun zero(): T
}

val <T : AbelianGroup<T>> AbelianGroup<T>.isZero: Boolean
    get() = this == zero()

operator fun <T : AbelianGroup<T>> AbelianGroup<T>.minus(t: T): T = this + -t
